File: src/app/core/PropertiesReader.java
package app.core;

import java.util.Properties;

import app.global.Items;
import app.global.Sources;
import app.rules.AllowedProperties;
import app.rules.DefProperties;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.HashMap;
import java.util.Map;

/**
 * @author Jack Meng
 */
public class PropertiesReader {
  private static Properties p;

  public PropertiesReader() {
    System.out.println("hasAll" + checkPropertiesAll() + "\nFile exists? " + new java.io.File(app.global.Items.items[1] + System.getProperty("file.separator") + app.global.Sources.PROPERTIES_FILE).exists());
    if (!new java.io.File(app.global.Items.items[1] + "/" + app.global.Sources.PROPERTIES_FILE).exists()) {
      reset();
    }
  }

  public Map<String, String> keyyedProp() throws IOException {
    HashMap<String, String> properties = new HashMap<>();
    p = new Properties();
    if (!new File(Items.items[1] + "/" + Sources.PROPERTIES_FILE).exists()) {
      reset();
    }
    try (InputStream isr = new FileInputStream(app.global.Items.items[1] + "/" + app.global.Sources.PROPERTIES_FILE)) {
      p.load(isr);
      if (AllowedProperties.valTheme(p.getProperty("gui.defaultTheme")))
        properties.put("gui.defaultTheme", p.getProperty("gui.defaultTheme"));

      if (AllowedProperties.valCache(p.getProperty("runner.disableCache")))
        properties.put("runner.disableCache", p.getProperty("runner.disableCache"));

      if (AllowedProperties.valBoxSize(p.getProperty("gui.defaultBoxSize")))
        properties.put("gui.defaultBoxSize", p.getProperty("gui.defaultBoxSize"));

      if (AllowedProperties.valBox(p.getProperty("gui.buttonShape"))) {
        properties.put("gui.buttonShape", p.getProperty("gui.buttonShape"));
      }

      if (AllowedProperties.valTransparency(p.getProperty("gui.window_transparency"))) {
        properties.put("gui.window_transparency", p.getProperty("gui.window_transparency"));
      }
    }
    return properties;
  }

  public boolean checkPropertiesAll() {
    p = new Properties();
    if(!new File(Items.items[1] + "/" + Sources.PROPERTIES_FILE).exists()) {
      return false;
    }
    try (InputStream isr = new FileInputStream(Items.items[1] + "/" + Sources.PROPERTIES_FILE)) {
      p.load(isr);

      if(!AllowedProperties.valTheme(p.getProperty("gui.defaultTheme")) || !AllowedProperties.valCache(p.getProperty("runner.disableCache")) || !AllowedProperties.valBoxSize(p.getProperty("gui.defaultBoxSize")) || !AllowedProperties.valBox(p.getProperty("gui.buttonShape")) || !AllowedProperties.valTransparency(p.getProperty("gui.window_transparency"))) {
        return false;
      }
    } catch (IOException e) {
      e.printStackTrace();
    }
    return true;
  } 

  public String setProperty(String key, String value) throws IOException {
    p.setProperty(key, value);
    try (OutputStream os = new FileOutputStream(
        new File(Items.items[1] + System.getProperty("file.separator") + Sources.PROPERTIES_FILE))) {
      p.store(os, Items.PROPERTIES_HEADER_COMMENT);
    }
    return value;
  }

  public static boolean reset() {
    try {
      new File(Items.items[1] + "/" + app.global.Sources.PROPERTIES_FILE).createNewFile();
    } catch (IOException e) {
      e.printStackTrace();
    }
    try (OutputStream os = new FileOutputStream(app.global.Items.items[1] + "/" + app.global.Sources.PROPERTIES_FILE)) {
      p = new java.util.Properties();
      p.setProperty("gui.defaultTheme", DefProperties.DEFAULT_GUI_LAF);
      p.setProperty("runner.disableCache", DefProperties.DISABLE_CACHE);
      p.setProperty("gui.defaultBoxSize", String.valueOf(DefProperties.DEFAULT_BOX_SIZE));
      p.setProperty("gui.buttonShape", DefProperties.DEFAULT_BUTTON_SHAPE);
      p.setProperty("gui.window_transparency", String.valueOf(DefProperties.DEFAULT_WINDOW_TRANSPARENCY));
      p.store(os, Items.PROPERTIES_HEADER_COMMENT);
    } catch (java.io.IOException e) {
      e.printStackTrace();
      return false;
    }
    return true;
  }

  public String getVal(String key) {
    return p.getProperty(key);
  }
}

File: src/app/core/PropertiesReader.kt
package app.core

import kotlin.Throws
import java.io.IOException
import java.util.HashMap
import app.core.PropertiesReader
import app.global.Items
import java.util.Properties
import app.global.Sources
import java.io.FileInputStream
import app.rules.AllowedProperties
import java.io.FileOutputStream
import app.rules.DefProperties
import java.io.File

/**
 * @author Jack Meng
 */
class PropertiesReader {
    @Throws(IOException::class)
    fun keyyedProp(): Map<String, String> {
        val properties = HashMap<String, String>()
        p = Properties()
        if (!File(Items.items[1].toString() + "/" + Sources.PROPERTIES_FILE).exists()) {
            reset()
        }
        FileInputStream(Items.items[1].toString() + "/" + Sources.PROPERTIES_FILE).use { isr ->
            p!!.load(isr)
            if (AllowedProperties.valTheme(p!!.getProperty("gui.defaultTheme"))) properties["gui.defaultTheme"] = p!!.getProperty("gui.defaultTheme")
            if (AllowedProperties.valCache(p!!.getProperty("runner.disableCache"))) properties["runner.disableCache"] = p!!.getProperty("runner.disableCache")
            if (AllowedProperties.valBoxSize(p!!.getProperty("gui.defaultBoxSize"))) properties["gui.defaultBoxSize"] = p!!.getProperty("gui.defaultBoxSize")
            if (AllowedProperties.valBox(p!!.getProperty("gui.buttonShape"))) {
                properties["gui.buttonShape"] = p!!.getProperty("gui.buttonShape")
            }
            if (AllowedProperties.valTransparency(p!!.getProperty("gui.window_transparency"))) {
                properties["gui.window_transparency"] = p!!.getProperty("gui.window_transparency")
            }
        }
        return properties
    }

    fun checkPropertiesAll(): Boolean {
        p = Properties()
        if (!File(Items.items[1].toString() + "/" + Sources.PROPERTIES_FILE).exists()) {
            return false
        }
        try {
            FileInputStream(Items.items[1].toString() + "/" + Sources.PROPERTIES_FILE).use { isr ->
                p!!.load(isr)
                if (!AllowedProperties.valTheme(p!!.getProperty("gui.defaultTheme")) || !AllowedProperties.valCache(p!!.getProperty("runner.disableCache")) || !AllowedProperties.valBoxSize(p!!.getProperty("gui.defaultBoxSize")) || !AllowedProperties.valBox(p!!.getProperty("gui.buttonShape")) || !AllowedProperties.valTransparency(p!!.getProperty("gui.window_transparency"))) {
                    return false
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return true
    }

    @Throws(IOException::class)
    fun setProperty(key: String?, value: String): String {
        p!!.setProperty(key, value)
        FileOutputStream(
                File(Items.items[1].toString() + System.getProperty("file.separator") + Sources.PROPERTIES_FILE)).use { os -> p!!.store(os, Items.PROPERTIES_HEADER_COMMENT) }
        return value
    }

    fun getVal(key: String?): String {
        return p!!.getProperty(key)
    }

    companion object {
        private var p: Properties? = null
        fun reset(): Boolean {
            try {
                val t = File(Items.items[1].toString() + "/" + Sources.PROPERTIES_FILE)
                if (!t.exists()) t.createNewFile()
            } catch (e: IOException) {
                e.printStackTrace()
            }
            try {
                FileOutputStream(Items.items[1].toString() + "/" + Sources.PROPERTIES_FILE).use { os ->
                    p = Properties()
                    p!!.setProperty("gui.defaultTheme", DefProperties.DEFAULT_GUI_LAF)
                    p!!.setProperty("runner.disableCache", DefProperties.DISABLE_CACHE)
                    p!!.setProperty("gui.defaultBoxSize", DefProperties.DEFAULT_BOX_SIZE.toString())
                    p!!.setProperty("gui.buttonShape", DefProperties.DEFAULT_BUTTON_SHAPE)
                    p!!.setProperty("gui.window_transparency", DefProperties.DEFAULT_WINDOW_TRANSPARENCY)
                    p!!.store(os, Items.PROPERTIES_HEADER_COMMENT)
                }
            } catch (e: IOException) {
                e.printStackTrace()
                return false
            }
            return true
        }
    }

    init {
        println("""
    hasAll${checkPropertiesAll()}
    File exists? ${File(Items.items[1].toString() + System.getProperty("file.separator") + Sources.PROPERTIES_FILE).exists()}
    """.trimIndent())
        if (!File(Items.items[1].toString() + "/" + Sources.PROPERTIES_FILE).exists()) {
            reset()
        }
    }
}